lib/auth/password.go

```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
 	return teleport.HOTP, nil
 }
 
-// GetOTPData returns the OTP Key, Key URL, and the QR code.
-func (s *AuthServer) GetOTPData(user string) (string, []byte, error) {
-	// get otp key from backend
-	otpSecret, err := s.GetTOTP(user)
+func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
+	clusterConfig, err := s.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if clusterConfig.GetLocalAuth() == false {
+		return nil, trace.AccessDenied(noLocalAuth)
+	}
+
+	err = services.VerifyPassword(req.Password)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	userToken, err := s.GetUserToken(req.TokenID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if userToken.Expiry().Before(s.clock.Now().UTC()) {
+		return nil, trace.BadParameter("expired token")
+	}
+
+	username := userToken.GetUser()
+	err = s.changeUserSecondFactor(req, userToken)
 	if err != nil {
-		return "", nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	// create otp url
-	params := map[string][]byte{"secret": []byte(otpSecret)}
-	otpURL := utils.GenerateOTPURL("totp", user, params)
+	err = s.deleteUserTokens(username)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	// create the qr code
-	otpQR, err := utils.GenerateQRCode(otpURL)
+	err = s.UpsertPassword(username, []byte(req.Password))
 	if err != nil {
-		return "", nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+
+	user, err := s.GetUser(username, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return user, nil
+}
+
+func (s *AuthServer) changeUserSecondFactor(req ChangePasswordWithTokenRequest, userToken services.UserToken) error {
+	username := userToken.GetUser()
+	cap, err := s.GetAuthPreference()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	switch cap.GetSecondFactor() {
+	case teleport.OFF:
+		return nil
+	case teleport.OTP, teleport.TOTP, teleport.HOTP:
+		secrets, err := s.Identity.GetUserTokenSecrets(req.TokenID)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// TODO: create a separate method to validate TOTP without inserting it first
```

Comment:
Is this TODO still valid, if not, please remove the comment.

Reply:
Yes, it is valid.